### ISSUE TITLE
clang does not warn about uninitialized reads in lambdas

### ISSUE THREAD

**Issue by nico** [llvm/llvm-project#128058]
This warns (as it should):

```
$ cat test.cc
struct S {};

void g() {
  S s1 = s1;
}
```

```
$ out/gn/bin/clang -c test.cc
test.cc:4:10: warning: variable 's1' is uninitialized when used within its own initialization [-Wuninitialized]
    4 |   S s1 = s1;
      |     ~~   ^~
1 warning generated.
```

This doesn't (but should, too):

```
$ cat test.cc
struct S {};

void g() {
  S s2 = [&]() { return s2; }();
}
```

```
$ out/gn/bin/clang -c test.cc
# No diag
```

**Comment by zhaohuiw42** [comment]
I'm interested in working on this. Can it be assigned to me?

**Comment by zyn0217** [comment]
@zhaohuiw42 Here it is. Though I'm not sure if this is a good-first-issue ...
